onnx/defs/reduction/utils.cc

```diff
@@ -0,0 +1,148 @@
+    if (axes_input) {
+      schema.Attr(
+          "noop_with_empty_axes",
+          "Defines behavior if 'axes' is empty. Default behavior with 'false' is to reduce all axes. "
+          "When axes is empty and this attribute is set to true, input tensor will not be reduced,"
+          "and the output tensor would be equivalent to input tensor.",
+          AttributeProto::INT,
+          static_cast<int64_t>(0));
+      schema.Input(
+          1,
+          "axes",
+          "Optional input list of integers, along which to reduce. "
+          "The default is to reduce over all the dimensions of the input tensor if 'noop_with_empty_axes' is false, "
+          "else act as an Identity op when 'noop_with_empty_axes' is true. "
+          "Accepted range is [-r, r-1] where r = rank(data).",
+          "tensor(int64)",
+          OpSchema::Optional,
+          true,
+          1,
+          OpSchema::NonDifferentiable);
+    } else {
+      schema.Attr(
+          "axes",
+          "A list of integers, along which to reduce. The default is to reduce over "
+          "all the dimensions of the input tensor. Accepted range is [-r, r-1] where r = rank(data).",
+          AttributeProto::INTS,
+          OPTIONAL_VALUE);
+    }
+    schema.Output(0, "reduced", "Reduced output tensor.", "T", OpSchema::Single, true, 1, OpSchema::Differentiable);
+    schema.TypeConstraint(
+        "T",
+        GetSupportedDataTypesForReductionOps(supports_8bit_datatypes),
+        supports_8bit_datatypes ? "Constrain input and output types to high-precision and 8 bit numeric tensors."
+                                : "Constrain input and output types to high-precision numeric tensors.");
+    if (func_body) {
+      schema.FunctionBody(func_body);
+    } else if (function_builder) {
+      std::cout << "schema.SetContextDependentFunctionBodyBuilder(function_builder);" << std::endl;
+      schema.SetContextDependentFunctionBodyBuilder(function_builder);
+    }
+    schema.TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+      propagateElemTypeFromInputToOutput(ctx, 0, 0);
+      if (!hasNInputShapes(ctx, 1)) {
+        return;
+      }
+
+      int64_t keep_dims = 1, noop_with_empty_axes = 0;
+      auto attr_proto = ctx.getAttribute("keepdims");
+      if (attr_proto) {
+        keep_dims = attr_proto->i();
+      }
+      auto noop_attr_proto = ctx.getAttribute("noop_with_empty_axes");
+      if (noop_attr_proto) {
+        noop_with_empty_axes = noop_attr_proto->i();
+      }
+      std::vector<int64_t> axes;
+      size_t num_inputs = ctx.getNumInputs();
+      if ((num_inputs == 2) && ctx.getInputType(1)) { // axes is input
+        if (ctx.getAttribute("axes")) {
+          fail_shape_inference("axes as an input and attribute cannot be specified at the same time.");
+        }
+
+        const TensorProto* axesInitializer = ctx.getInputData(1);
+        if (axesInitializer == nullptr) {
+          // skip if axes is not an initializer
```

Comment:
it makes sense to check the rank of the axes input which shall be 1. I will update this.
Constants is converted to initializer by the runtime. Here I should use axes_constant as the name of the variable. I will update this PR for this.

ONNX shape inference for existing cases (where axes is just constant numbers) is handled in this PR. In future cases where constant folding can help, I believe ONNX shape inference shall be enhanced to do that.

Reply:
> Dynamic axes input potentially introduces dynamic output rank for reduce ops. However it is likely the axes values in most "new" cases can be either constant folded computed or symbolically inferred. @jcwchen any thoughts?

Good catch. Yes, if the "axes" becomes input, it's possible that in the future such a shape information needs to be dynamic computed. Currently reduce-family operators does not support symbolic shape inference yet. I think for now it should be fine because this PR promotes axes from attribute to input for generic reason, instead of having a real-world model requires the axes to be dynamic computed. After promoting axes to input, if the shape information comes from either a Constant op or initializer, I think current shape inference can still handle that.

Still, I agreed that supporting dynamic input for reduce-family operator is definitely good-to-have. A question for @liqunfu: in the future will reduction operators involve dynamic computing for axes when using it in a function op?

Reply:
It is hard to say for sure. It can only be done if the input come from a constant op. In other cases, axes cannot be calculated.  However in such cases the existing reduce ops cannot work either.